applications/nrf_desktop/src/util/config_channel_transport.c

```diff
@@ -0,0 +1,272 @@
+		return -ENOTSUP;
+	}
+
+	__ASSERT_NO_MSG(event->dyndata.size >= config_data_len);
+	event->dyndata.size = config_data_len;
+
+	memcpy(event->dyndata.data, &buffer[pos], config_data_len);
+	pos += config_data_len;
+
+	return pos;
+}
+
+int config_channel_report_fill(uint8_t *buffer, const size_t length,
+			       const struct config_event *event)
+{
+	int err = frame_length_check(length);
+	if (!err) {
+		err = data_len_check(event->dyndata.size);
+	}
+
+	if (err) {
+		return err;
+	}
+
+	size_t pos = 0;
+
+	sys_put_le16(event->recipient, &buffer[pos]);
+	pos += sizeof(event->recipient);
+
+	buffer[pos] = event->event_id;
+	pos += sizeof(event->event_id);
+
+	buffer[pos] = event->status;
+	pos += sizeof(event->status);
+
+	buffer[pos] = (uint8_t)event->dyndata.size;
+	pos += sizeof(uint8_t); /* Data length is encoded using one byte. */
+
+	__ASSERT_NO_MSG(pos == TRANSPORT_HEADER_SIZE);
+	__ASSERT_NO_MSG(pos + event->dyndata.size <= length);
+
+	if (event->dyndata.size > 0) {
+		memcpy(&buffer[pos], event->dyndata.data, event->dyndata.size);
+		pos += event->dyndata.size;
+	}
+
+	return pos;
+}
+
+static void drop_transactions(struct config_channel_transport *transport)
+{
+	/* Transport ID is changed, responses for submitted requests will
+	 * be dropped.
+	 */
+	uint8_t *new_ptr = transport->transport_id;
+
+	new_ptr++;
+	/* Make sure we will not overlap with another transport. */
+	if ((new_ptr - (uint8_t *)transport) >= sizeof(*transport)) {
+		new_ptr = (uint8_t *)transport;
+	}
+
+	transport->transport_id = new_ptr;
+}
+
+static void timeout_fn(struct k_work *work)
+{
+	struct config_channel_transport *transport = CONTAINER_OF(work,
+					struct config_channel_transport,
+					timeout);
+
+	__ASSERT_NO_MSG(transport->state == CONFIG_CHANNEL_TRANSPORT_WAIT_RSP);
+
+	/* Send response with timeout status, without aditional data. */
+	transport->data[3] = CONFIG_STATUS_TIMEOUT;
+	drop_transactions(transport);
+	transport->state = CONFIG_CHANNEL_TRANSPORT_RSP_READY;
+}
+
+void config_channel_transport_init(struct config_channel_transport *transport)
+{
+	__ASSERT_NO_MSG(transport->state == CONFIG_CHANNEL_TRANSPORT_DISABLED);
+	transport->state = CONFIG_CHANNEL_TRANSPORT_IDLE;
+	transport->transport_id = transport;
+	k_delayed_work_init(&transport->timeout, timeout_fn);
+}
+
+int config_channel_transport_get(struct config_channel_transport *transport,
+				 uint8_t *buffer, size_t length)
+{
+	__ASSERT_NO_MSG(transport->state != CONFIG_CHANNEL_TRANSPORT_DISABLED);
+
+	if (length < transport->data_len) {
+		LOG_ERR("Host fetched incomplete data");
+	}
+
+	memcpy(buffer, transport->data, length);
+
+	if (transport->state == CONFIG_CHANNEL_TRANSPORT_RSP_READY) {
+		transport->state = CONFIG_CHANNEL_TRANSPORT_IDLE;
+	}
+
+	return 0;
+}
+
+int config_channel_transport_set(struct config_channel_transport *transport,
+				 const uint8_t *buffer, size_t length)
+{
+	__ASSERT_NO_MSG(transport->state != CONFIG_CHANNEL_TRANSPORT_DISABLED);
```

Comment:
Since data come from the outside world maybe we should handle state checking correctly?

Reply:
Assertions are added, because we assume that the config_channel_transport will be initialized by a module (hids/usb_state) before it starts receiving the data (we need to submit work related to timeout when config channel transaction starts). Also we assume that timeout will not be called in case we are not waiting for response.

These asserts actually rely on internal data, not on outside world.